File: rules/go/tests/go.path-traversal/path.go
package main

import (
	"net/http"
	"os"
	"path/filepath"
)

func read(r *http.Request) {
	os.ReadFile("/data/" + r.FormValue("name")) // crit:expect go.path-traversal
}

func viaVar(r *http.Request) {
	name := r.URL.Query().Get("name")
	// crit:expect go.path-traversal
	os.Open("/data/" + name)
}

func safe(r *http.Request) {
	os.ReadFile(filepath.Base(r.FormValue("name"))) // crit:expect-not go.path-traversal
}

func fixed() {
	os.ReadFile("/etc/app.conf") // crit:expect-not go.path-traversal
}

File: rules/csharp/tests/cs.path-traversal/Path.cs
class PathTests {
  void Read() {
    var name = Request.QueryString["name"];
    File.ReadAllText("/data/" + name); // crit:expect cs.path-traversal
  }
  void Write() {
    File.WriteAllText(Request.Form["path"], "x"); // crit:expect cs.path-traversal
  }
  void Safe() {
    var name = Request.QueryString["name"];
    File.ReadAllText(Path.GetFileName(name)); // crit:expect-not cs.path-traversal
  }
  void Fixed() {
    File.ReadAllText("/etc/app.config"); // crit:expect-not cs.path-traversal
  }
}
